File: baseline/MacHost/Sources/VideoColorNegotiation.swift
import Foundation
import VibeScreenProtocol

enum HostVideoColorDecision: Equatable {
    case accepted(VSColorDescription)
    case fallback(VSColorDescription, reason: String)
    case rejected(reason: String)
}

struct HostVideoColorNegotiator {
    static let unsupportedHDRFallbackReason = "unsupported_color_or_decode_profile"

    let clientCapabilities: Set<VSCapability>
    let decodeCapabilities: [VSVideoDecodeCapability]

    func evaluate(
        _ requested: VSColorDescription,
        codec: VSCodec,
        encodedSize: VSDimensions,
        framesPerSecond: UInt32
    ) -> HostVideoColorDecision {
        let color = Self.normalized(requested)
        let supportsLegacySDR = supports(
            Self.legacySDRColor,
            codec: codec,
            encodedSize: encodedSize,
            framesPerSecond: framesPerSecond
        )
        if Self.isHDR(color), !clientCapabilities.isSuperset(of: [.colorManagement, .hdrVideo]) {
            return unsupportedDecision(supportsLegacySDR: supportsLegacySDR)
        }
        guard supports(color, codec: codec, encodedSize: encodedSize, framesPerSecond: framesPerSecond) else {
            return unsupportedDecision(supportsLegacySDR: supportsLegacySDR)
        }
        return .accepted(color)
    }

    func validatedFallback(
        _ selected: VSColorDescription,
        codec: VSCodec,
        encodedSize: VSDimensions,
        framesPerSecond: UInt32
    ) -> VSColorDescription? {
        let color = Self.normalized(selected)
        guard !Self.isHDR(color), supports(
            color,
            codec: codec,
            encodedSize: encodedSize,
            framesPerSecond: framesPerSecond
        ) else { return nil }
        return color
    }

    func supportsLegacySDR(
        codec: VSCodec,
        encodedSize: VSDimensions,
        framesPerSecond: UInt32
    ) -> Bool {
        supports(
            Self.legacySDRColor,
            codec: codec,
            encodedSize: encodedSize,
            framesPerSecond: framesPerSecond
        )
    }

    static var legacySDRColor: VSColorDescription {
        var color = VSColorDescription()
        color.primaries = .bt709
        color.transferFunction = .bt709
        color.matrixCoefficients = .bt709
        color.fullRange = false
        color.bitDepth = 8
        return color
    }

    static func normalized(_ color: VSColorDescription) -> VSColorDescription {
        var normalized = color
        if normalized.primaries == .unspecified { normalized.primaries = .bt709 }
        if normalized.transferFunction == .unspecified { normalized.transferFunction = .bt709 }
        if normalized.matrixCoefficients == .unspecified { normalized.matrixCoefficients = .bt709 }
        if normalized.bitDepth == 0 { normalized.bitDepth = 8 }
        return normalized
    }

    private func supports(
        _ color: VSColorDescription,
        codec: VSCodec,
        encodedSize: VSDimensions,
        framesPerSecond: UInt32
    ) -> Bool {
        guard !decodeCapabilities.isEmpty else { return !Self.isHDR(color) }
        return decodeCapabilities.contains(where: { capability in
            let transfer = color.transferFunction == .unspecified ? .bt709 : color.transferFunction
            let sizeMatches = encodedSize.width <= capability.maximumWidth
                && encodedSize.height <= capability.maximumHeight
            let frameRateMatches = framesPerSecond <= capability.maximumFramesPerSecond
            return capability.codec == codec
                && sizeMatches
                && frameRateMatches
                && capability.bitDepths.contains(color.bitDepth == 0 ? 8 : color.bitDepth)
                && capability.transferFunctions.contains(transfer)
        })
    }

    private func unsupportedDecision(supportsLegacySDR: Bool) -> HostVideoColorDecision {
        if supportsLegacySDR {
            return .fallback(Self.legacySDRColor, reason: Self.unsupportedHDRFallbackReason)
        }
        return .rejected(reason: Self.unsupportedHDRFallbackReason)
    }

    static func isHDR(_ color: VSColorDescription) -> Bool {
        color.bitDepth > 8
            || color.transferFunction == .pq
            || color.transferFunction == .hlg
            || color.primaries == .bt2020
            || color.matrixCoefficients == .bt2020NonConstant
    }
}

File: baseline/AndroidClient/app/src/main/java/dev/telemachus/display/protocol/VideoColorNegotiation.kt
package dev.telemachus.display.protocol

import dev.vibescreen.protocol.v1.ColorDescription
import dev.vibescreen.protocol.v1.ColorPrimaries
import dev.vibescreen.protocol.v1.MatrixCoefficients
import dev.vibescreen.protocol.v1.TransferFunction
import dev.vibescreen.protocol.v1.VideoDecodeCapability

internal sealed class VideoColorDecision {
    object Accepted : VideoColorDecision()

    data class Fallback(
        val selectedColor: ColorDescription,
        val reason: String,
    ) : VideoColorDecision()

    data class Rejected(
        val reason: String,
    ) : VideoColorDecision()
}

internal object VideoColorNegotiation {
    const val UNSUPPORTED_COLOR_OR_DECODE_PROFILE = "unsupported_color_or_decode_profile"

    val legacySdrColor: ColorDescription =
        ColorDescription
            .newBuilder()
            .setPrimaries(ColorPrimaries.COLOR_PRIMARIES_BT709)
            .setTransferFunction(TransferFunction.TRANSFER_FUNCTION_BT709)
            .setMatrixCoefficients(MatrixCoefficients.MATRIX_COEFFICIENTS_BT709)
            .setFullRange(false)
            .setBitDepth(8)
            .build()

    fun sdrDecodeCapabilities(codecs: Iterable<dev.vibescreen.protocol.v1.Codec>): List<VideoDecodeCapability> =
        codecs.map { codec ->
            VideoDecodeCapability
                .newBuilder()
                .setCodec(codec)
                .setMaximumWidth(MAXIMUM_SDR_WIDTH)
                .setMaximumHeight(MAXIMUM_SDR_HEIGHT)
                .setMaximumFramesPerSecond(MAXIMUM_SDR_FRAMES_PER_SECOND)
                .addBitDepths(8)
                .addTransferFunctions(TransferFunction.TRANSFER_FUNCTION_BT709)
                .addTransferFunctions(TransferFunction.TRANSFER_FUNCTION_SRGB)
                .build()
        }

    fun evaluate(
        requestedColor: ColorDescription?,
        negotiatedHdr: Boolean,
        decodeCapabilities: Iterable<VideoDecodeCapability>,
        codec: dev.vibescreen.protocol.v1.Codec,
        width: Int,
        height: Int,
        framesPerSecond: Int,
    ): VideoColorDecision {
        val color = normalize(requestedColor ?: legacySdrColor)
        val supportsLegacySdr = supports(decodeCapabilities, legacySdrColor, codec, width, height, framesPerSecond)
        if (isHdr(color) && !negotiatedHdr) {
            return unsupportedDecision(supportsLegacySdr)
        }
        val supported = supports(decodeCapabilities, color, codec, width, height, framesPerSecond)
        return if (supported) {
            VideoColorDecision.Accepted
        } else {
            unsupportedDecision(supportsLegacySdr)
        }
    }

    fun normalize(color: ColorDescription): ColorDescription {
        val builder = color.toBuilder()
        if (builder.primaries == ColorPrimaries.COLOR_PRIMARIES_UNSPECIFIED) {
            builder.primaries = ColorPrimaries.COLOR_PRIMARIES_BT709
        }
        if (builder.transferFunction == TransferFunction.TRANSFER_FUNCTION_UNSPECIFIED) {
            builder.transferFunction = TransferFunction.TRANSFER_FUNCTION_BT709
        }
        if (builder.matrixCoefficients == MatrixCoefficients.MATRIX_COEFFICIENTS_UNSPECIFIED) {
            builder.matrixCoefficients = MatrixCoefficients.MATRIX_COEFFICIENTS_BT709
        }
        if (builder.bitDepth == 0) {
            builder.bitDepth = 8
        }
        return builder.build()
    }

    private fun isHdr(color: ColorDescription): Boolean =
        color.bitDepth > 8 ||
            color.transferFunction == TransferFunction.TRANSFER_FUNCTION_PQ ||
            color.transferFunction == TransferFunction.TRANSFER_FUNCTION_HLG ||
            color.primaries == ColorPrimaries.COLOR_PRIMARIES_BT2020 ||
            color.matrixCoefficients == MatrixCoefficients.MATRIX_COEFFICIENTS_BT2020_NON_CONSTANT

    private fun unsupportedDecision(supportsLegacySdr: Boolean): VideoColorDecision =
        if (supportsLegacySdr) {
            VideoColorDecision.Fallback(legacySdrColor, UNSUPPORTED_COLOR_OR_DECODE_PROFILE)
        } else {
            VideoColorDecision.Rejected(UNSUPPORTED_COLOR_OR_DECODE_PROFILE)
        }

    private fun supports(
        decodeCapabilities: Iterable<VideoDecodeCapability>,
        color: ColorDescription,
        codec: dev.vibescreen.protocol.v1.Codec,
        width: Int,
        height: Int,
        framesPerSecond: Int,
    ): Boolean =
        decodeCapabilities.any { capability ->
            capability.codec == codec &&
                width <= capability.maximumWidth &&
                height <= capability.maximumHeight &&
                framesPerSecond <= capability.maximumFramesPerSecond &&
                capability.bitDepthsList.contains(color.bitDepth) &&
                capability.transferFunctionsList.contains(color.transferFunction)
        }

    private const val MAXIMUM_SDR_WIDTH = 3840
    private const val MAXIMUM_SDR_HEIGHT = 2160
    private const val MAXIMUM_SDR_FRAMES_PER_SECOND = 120
}
